File: problems/2390/Lc2390.java
public class Lc2390 {
  static Lc2390 leetcode = new Lc2390();
  static Solution sol = leetcode.new Solution();

  class Solution {
    public String removeStars(String s) {
      char[] buf = s.toCharArray();
      int bufSize = 0;
      for (int pos = 0, len = buf.length; pos < len; pos++) {
        if (buf[pos] == '*') {
          if (bufSize > 0)
            --bufSize;
        } else {
          buf[bufSize++] = buf[pos];
        }
      }
      return new String(buf, 0, bufSize);
    }
  }

  static void testcase01() {
    String s = "leet**cod*e";
    String result = sol.removeStars(s);
    String expected = "lecoe";
    assert expected.equals(result) : "expected" + String.valueOf(expected) + ", but " + String.valueOf(result);
  }

  static void testcase02() {
    String s = "erase*****";
    String result = sol.removeStars(s);
    String expected = "";
    assert expected.equals(result) : "expected" + String.valueOf(expected) + ", but " + String.valueOf(result);
  }

  static void testcase03() {
    String s = "erase";
    String result = sol.removeStars(s);
    String expected = "erase";
    assert expected.equals(result) : "expected" + String.valueOf(expected) + ", but " + String.valueOf(result);
  }

  public static void main(String[] args) {
    testcase01();
    testcase02();
    testcase03();
  }
}

File: problems/2390/Lc2390.c
#include <assert.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>

char* removeStars(char* s) {
  size_t size = strlen(s), lpos = 0;
  char* buf = (char*)malloc((size + 1) * sizeof(char));
  for (size_t rpos = 0; rpos < size; ++rpos) {
    if (s[rpos] == '*') {
      if (lpos > 0)
        --lpos;
    } else {
      buf[lpos++] = s[rpos];
    }
  }
  buf[lpos] = '\0';
  return buf;
}

void testcase01() {
  char* s = "leet**cod*e";
  char* result = removeStars(s);
  char* expected = "lecoe";
  assert(strcmp(expected, result) == 0);
}

void testcase02() {
  char* s = "erase*****";
  char* result = removeStars(s);
  char* expected = "";
  assert(strcmp(expected, result) == 0);
}

void testcase03() {
  char* s = "erase";
  char* result = removeStars(s);
  char* expected = "erase";
  assert(strcmp(expected, result) == 0);
}

int main() {
  testcase01();
  testcase02();
  testcase03();
  return 0;
}
